Allow f(t, **args) as coefficient function alternative (alternative approach)

hodgestar opened the pull request:
**Description**
Allow function-based Coefficients and QobjEvo elements to support functions with more natural argument signatures like `f(t, w)` or `f(t, **kw)` in addition to the old `f(t, args)` signature. The old style signature is used if the arguments to the function are exactly `(t, args)`. Otherwise the new calling convention is used.

This PR also limits the arguments stored to those accepted by the function, potentially reducing some extra copies of functions when `.replace_arguments()` is called.

**Related issues or PRs**
This is an alternative implementation of #1609. This implementation retains single FunctionCoefficient and _FuncElement classes and simplifies the logic for deciding whether the old or new-style calling convention is used.

**Changelog**
Allow function-based Coefficients and QobjEvo elements to support functions with more natural argument signatures like `f(t, w)` or `f(t, **kw)` in addition to the old `f(t, args)`.

Ericgig commented:
It's very clean, I really like the `set & set` and reusing the classes.

But in v4, while we used `f(t, args)` in most examples, we never said that it should be the exact signature up to variable name, so I don't think it an appropriate check for which version to use. Ideally, the check should be good enough we don't need to change any tests not directly related to the addition of this feature.

Also what should
```
def f(t, w, a=1):
    return a

coeff = coefficient(f, args={'w':1})
coeff.replace_arguments(a=2)(0)
```
give? Since `a` is not given, it's not our job to update it, or should we add it to `_f_arg_names`?

hodgestar commented:
> It's very clean, I really like the `set & set` and reusing the classes.

Thanks!

> But in v4, while we used `f(t, args)` in most examples, we never said that it should be the exact signature up to variable name, so I don't think it an appropriate check for which version to use. Ideally, the check should be good enough we don't need to change any tests not directly related to the addition of this feature.

I'm open to modifying the check for QuTiP v4 functions, but I would like to keep the check simple. To quote the Zen of Python, "If the implementation is hard to explain, it's a bad idea.". :)

> Also what should
> 
> ```
> def f(t, w, a=1):
>     return a
> 
> coeff = coefficient(f, args={'w':1})
> coeff.replace_arguments(a=2)(0)
> ```
> 
> give? Since `a` is not given, it's not our job to update it, or should we add it to `_f_arg_names`?

I think this works fine already in this branch. `a` is included in `_f_arg_names`, so it can be specified via `.replace_arguments(...)` but since it has a default it can also be left out.

hodgestar commented:
@Ericgig I'm back from vacation so I'm keen to restart the discussion on this and hopefully get it merge. Any suggestions for a better way to tell whether the coefficient function expects the old-style arguments? I'm happy to do the coding, but I'd like to keep the check simple and I don't have a better idea for it right now.

hodgestar commented:
@Ericgig I've implemented the setting we discussed and removed the `by_inspection` additional constructor (and replaced it by just having some internal use only parameters in the constructors).

Ready for re-review.

Ericgig commented on qutip/core/cy/coefficient.pyx:
```diff
@@ -16,6 +17,62 @@ cdef extern from "<complex>" namespace "std" nogil:
+    if style is None:
+        style = qutip.settings.core["function_coefficient_style"]
+    if style == "auto":
+        if tuple(sig.parameters.keys()) == ("t", "args") and not f_has_kw:
+            # if the signature is exactly f(t, args), then assume parameters
+            # are supplied in an argument dictionary
+            style = "dict"
```
We never said to the user to use this exact signature and adding it now calling it _backward compatibility_ feels wrong. We can do better.
_Function that if called with `f(t, **args)` would be missing the second positional argument._ would be a good condition.
These functions would fail if called as pythonic so we are still giving it more weight, it's more inclusive than a strict signature without being overly greedy. This would support `f(t, _)` and `f(t, kwargs)` which are other natural signature for the dict format. My expectation is that for most user not using decorator, this would work as expected. Users using `partial` etc. will need to update their code, but the `function_style` is there for them.
 
```suggestion
        keys = tuple(sig.parameters.keys())
        if (
            len(keys) == 2
            and not f_has_kw
            and keys[1] not in args
            and sig.parameters[keys[1]].default is inspect._empty
        ):
```

Ericgig replied:
My understanding was there would be no `style='auto'`, just `'dict'` or `'pythonic'`. So I saw `'auto'` as a try at backward compatibility, for which it's not great, but as a way to support both v4 and v5 by projects using qutip, it's look fine. These projects would use `f(t, args)` to support both version until they start using other new features and drop v4, but that is still far ahead.

`qutip.settings.core`'s options will be useful for users trying out v5. My guess is that many users will try both version for their research and take the fastest if they don't need the new features. (openmp will probably not be working on v5 in the beginning, etc.) So I want as little friction in passing from v4 to v5 for normal qutip usage.

About point 4, we should add context with options, while not needed here with this solution, it would be nice for testing global options if nothing else. But it would allow project to use their own options if needed.

hodgestar commented on qutip/core/cy/_element.pyx:
```diff
@@ -356,30 +357,72 @@ cdef class _EvoElement(_BaseElement):
         self._previous = (Nan, None)
 
+    @classmethod
+    def by_inspection(cls, func, args):
```
Done!

hodgestar commented on qutip/core/cy/_element.pyx:
```diff
@@ -356,30 +357,72 @@ cdef class _EvoElement(_BaseElement):
+    def by_inspection(cls, func, args):
+        sig = inspect.signature(func)
+        f_is_kw = any(p.kind == p.VAR_KEYWORD for p in sig.parameters.values())
+        f_is_t_args = tuple(sig.parameters.keys()) == ("t", "args") and not f_is_kw
```
We have a new discussion about this on the more recent commits, so resolving this older comment.

hodgestar commented on qutip/core/cy/_element.pyx:
```diff
@@ -356,28 +357,67 @@ cdef class _EvoElement(_BaseElement):
     """
-    def __init__(self, func, args):
+
+    _UNSET = object()
+
+    def __init__(self, func, args, style=None, bint _f_pythonic=_UNSET, _f_parameters=_UNSET):
+        if _f_pythonic is self._UNSET or _f_parameters is self._UNSET:
+            _f_pythonic, _f_parameters = coefficient_function_parameters(
+                func, style=style)
```
I added `_UNSET` because `_f_parameters=None` has the special meaning "all parameters accepted".

I used `_UNSET` for `_f_pythonic` too for consistency with `_f_parameters` because they are used together.

I forgot to remove the `bint` in the function signature here -- thank you for catching that.

I think the extra type check for when only one is _UNSET is a good idea, but I will rework it a little so that only one check happens on the fast-path through the constructor.

hodgestar replied:
Implemented.

hodgestar commented on qutip/core/cy/_element.pxd:
```diff
@@ -25,6 +25,8 @@ cdef class _FuncElement(_BaseElement):
     cdef object _func
     cdef dict _args
     cdef tuple _previous
+    cdef object _f_pythonic
+    cdef object _f_parameters
```
I made them `object` so that I could pass `_UNSET`, but we never store `_UNSET` so I can put them back.

hodgestar replied:
I have putted the types back.

hodgestar commented on qutip/core/cy/coefficient.pyx:
```diff
@@ -112,24 +169,51 @@ cdef class FunctionCoefficient(Coefficient):
     """
     cdef object func
+    cdef object _f_pythonic
+    cdef object _f_parameters
```
Types added.

hodgestar commented on qutip/core/cy/_element.pyx:
```diff
@@ -411,13 +454,22 @@ cdef class _FuncElement(_BaseElement):
-                return new
-        new = _FuncElement(self._func, {**self._args, **args})
-        cache.append((self, new))
+        if self._f_parameters is not None:
+            args = {k: args[k] for k in self._f_parameters & args.keys()}
```
Removing the extra copying of args in `__init__` is a good idea -- I think that's partly left over from before the coefficient and element implementations were immutable.

I left in `.copy()` but only when `_f_pythonic` and `_f_parameters` are not supplied (i.e. when the args dictionary is passed in from outside and might be modified later by some other code).

Ericgig commented:
I see the tests for the last commit as both passing and failing...
Could you merge dev.major to rerun the tests.

hodgestar commented:
> I see the tests for the last commit as both passing and failing...
> Could you merge dev.major to rerun the tests.

Merged.

Ericgig reviewed (approved):
Look ready to merge.
It would be nice to rebase to remove commits from August which are no longer relevant.

hodgestar commented:
Rebase done. Just waiting for tests to finish re-running.
